typing/typedecl.ml

```diff
@@ -1780,22 +1896,27 @@ let update_decl_jkind env dpath decl =
 
   (* check that the jkind computed from the kind matches the jkind
      annotation, which was stored in decl.type_jkind *)
-  if new_jkind != decl.type_jkind then
+  if new_jkind != decl.type_jkind then begin
     (* CR layouts v2.8: Consider making a function that doesn't compute
        histories for this use-case, which doesn't need it. *)
-    begin match Jkind.sub_jkind_l new_jkind decl.type_jkind with
+    (* The [decl.type_jkind] should never have [with]-types in it at this
+       point. *)
```

Comment:
I suspect this is either wrong, or broken, and that's what's causing the failure in https://github.com/ocaml-flambda/flambda-backend/pull/3401#discussion_r1896882989

Reply:
in fact, if I replace this constantly-false `type_equal` with `Ctype.type_equal env`, a bunch of tests marked with "CR something: this should be accepted" or the like start being accepted. So I suspect that this should be actually using type equality. That said, what was the original idea behind this comment? I dug around and I couldn't figure out anything that might've been trying to strip baggage off the type_jkind of the decl, and in fact when constructing the type_jkind of the decl we explicitly add baggage from fields

Reply:
I've convinced myself enough that this was the right thing to do that I've pushed c03d79c09b4b93984ef60212dc765938345a7d63 making the change. feel free to revert if there's some subtlety I'm missing! but also the few branches I have based on this branch benefit from that change as well.